src/uct/api/uct_def.h

```diff
@@ -177,18 +200,47 @@ typedef struct uct_cm_ep_priv_data_pack_args {
      * @ref uct_cm_ep_priv_data_pack_args_field.
      * Fields not specified by this mask should not be accessed by the callback.
      */
-    uint64_t                   field_mask;
+    uint64_t                    field_mask;
 
     /**
      * Device name. This routine may fill the user's private data according to
      * the given device name. The device name that is passed to this routine,
      * corresponds to @ref uct_tl_resource_desc_t::dev_name as returned from
      * @ref uct_md_query_tl_resources.
      */
-    char                       dev_name[UCT_DEVICE_NAME_MAX];
+    char                        dev_name[UCT_DEVICE_NAME_MAX];
 } uct_cm_ep_priv_data_pack_args_t;
 
 
+/**
+ * @ingroup UCT_CLIENT_SERVER
+ * @brief Arguments to the client-server resolved callback.
+ *
+ * Used with the client-server API on a connection manager.
+ */
+typedef struct uct_cm_ep_resolve_args {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_cm_ep_priv_data_pack_args_field.
+     * Fields not specified by this mask should not be accessed by the callback.
+     */
+    uint64_t                    field_mask;
+
+    /**
+     * The endpoint which was bound to local device.
+     */
+    uct_ep_h                    ep;
```

Comment:
Is EP is really optional here or always is set ?

Reply:
in fact it is expected to be set always but may depend on implementation or be changed in future for example

Reply:
Can you please clarify why  this depends on implementation, especially when it is expect to be alway set ? We had similar discussion before, if this expect to be always set, it should go as a regular arg.

Reply:
`field_mask` and corresponding field are filled by caller i.e. implementation is responsible for this.
upper layer can decide to use this CM or fallback to other one, for example.

if we push all expected (today) fields to arg list we can get another not compatible API
in UCT we have the following conversion for callbacks:
```
status_t foo(void* arg, extra_params_t *params);
```
where `arg` is an argument provided by user, `params` is a forward compatible structure with args provided by transport

Reply:
Since this endpoint centric API, I don't see how ever it will be optional. This is like making "tag" or distention ep optional in send API.

Reply:
i tend to agree. @evgeny-leksikov do you think it's possible this callback will be called without an ep?

Reply:
> do you think it's possible this callback will be called without an ep?

in my current implementation - no, but it can be optional for user for example, let's say using user_data + his tricky data structure to access the ep.

please let me provide an example - just couple days ago we discussed (internaly) the case where we could ignore "device_name" in this resolve calback which was initially supposed to be used exactly to determinate the device (!). Tomorow we may have a need to make it optional because of some new transport limitation or to add another field to the structure which could contradict to this particular one. Example from the past - `uct_ep_create` which was initially designed to be "iface centric".

I would avoid adding many mandatory (today) args to the regular arg list, so I would suggest to return back to initial proposal:
```
typedef ucs_status_t (*uct_cm_ep_resolve_callback_t)(uct_ep_h ep, uct_cm_ep_resolve_args_t *resolve_args);
```
i.e. swap `user_data` and `ep` and make it ep-centric or follow common callback convention like it is now (+ probably drop "ep" from the type names)

what are yours thoughts?

Reply:
to which object's constructor do we pass this callback? if this is an ep, seems like it's mandatory anyway, no?

Reply:
@yosefe you know my opinion on this :) just an exmple from UCP:
https://github.com/openucx/ucx/blob/f1c7258aa91a52b1a538b8f4ad0890832e68b778/src/ucp/api/ucp.h#L2796
and its callback
https://github.com/openucx/ucx/blob/f1c7258aa91a52b1a538b8f4ad0890832e68b778/src/ucp/api/ucp_def.h#L611

Reply:
so why do we have a "uct_ep ep" field at all? let's just have optional user defined "void *arg"

Reply:
because it's for free for UCT and useful for UCP.
there is a small difference against AM, the EP is not created yet before `uct_ep_create` call, so we can not just pass not existing object to user_data arg, it could requre one more level of indirection in UCP to access the UCT EP from the callback.
(BTW I voted to add worker to AM callback)

Reply:
removed since it's easy to add when needed

Reply:
@shamisp we came to the conclusion this field is optional, and even more so, we don't really need it today (e.g we can have a working UCP without it). so better just to remove.

Reply:
@yosefe There is also EP in the name of the function ... what does exactly do we resolve here ?

Reply:
The ep in the function name means the callback is assigned for an endpoint.
Same as with active message handlers: we have "iface" in the function name, but not as an argument to the callback:
https://github.com/openucx/ucx/blob/db188f40865da1957b5b219d85d985bf221b0aad/src/uct/base/uct_iface.c#L83

Reply:
@yosefe make sense. Since I don't see how the new API is exactly integrated I want to ask about device name lookup ? How the device name get's looked up on UCP level ?

Reply:
@shamisp see https://github.com/openucx/ucx/blob/f1c7258aa91a52b1a538b8f4ad0890832e68b778/src/ucp/core/ucp_context.c#L1727
usage in current implementation
on client side:
https://github.com/openucx/ucx/blob/51d8eaee6437ac09ac6f5f11933bf8bb46cf16d6/src/ucp/wireup/wireup_cm.c#L129
on server side:
https://github.com/openucx/ucx/blob/51d8eaee6437ac09ac6f5f11933bf8bb46cf16d6/src/ucp/wireup/wireup_cm.c#L1019

Reply:
@evgeny-leksikov this helps. I'm taking a look ....